kinto/plugins/accounts/authentication.py

```diff
@@ -34,13 +41,62 @@ def account_check(username, password, request):
     except storage_exceptions.ObjectNotFoundError:
         return None
 
+    if validation_enabled and not is_validated(existing):
+        return None
+
     hashed = existing["password"].encode(encoding="utf-8")
     pwd_str = password.encode(encoding="utf-8")
     # Check if password is valid (it is a very expensive computation)
     if bcrypt.checkpw(pwd_str, hashed):
         cache.set(cache_key, hashed_password, ttl=cache_ttl)
         return True
 
+    # Last chance, is this a "reset password" flow?
+    reset_password_flow(username, password, request)
+
+
+def reset_password_flow(username, password, request):
+    settings = request.registry.settings
+    hmac_secret = settings["userid_hmac_secret"]
+    cache_key = utils.hmac_digest(hmac_secret, ACCOUNT_CACHE_KEY.format(username))
+    hashed_password = utils.hmac_digest(cache_key, password)
+    pwd_str = password.encode(encoding="utf-8")
+    cache_ttl = int(settings.get("account_cache_ttl_seconds", 30))
+    cache = request.registry.cache
+
+    cached_password = get_cached_reset_password(username, request.registry)
+    if not cached_password:
+        return None
+
+    # The temporary reset password is only available for changing a user's password.
+    if request.method.lower() not in ["post", "put", "patch"]:
+        return None
+
+    # Only allow modifying a user account, no other resource.
+    uri = utils.strip_uri_prefix(request.path)
+    resource_name, _ = utils.view_lookup(request, uri)
+    if resource_name != "account":
+        return None
+
+    try:
+        data = request.json["data"]
+    except ValueError:
+        return None
+
+    # Request one and only one data field: the `password`.
+    if not data or "password" not in data or len(data.keys()) > 1:
+        return None
+
+    cached_password_str = cached_password.encode(encoding="utf-8")
+    if bcrypt.checkpw(pwd_str, cached_password_str):
+        # Remove the temporary reset password from the cache.
+        reset_password_cache_key = utils.hmac_digest(
+            hmac_secret, ACCOUNT_RESET_PASSWORD_CACHE_KEY.format(username)
+        )
+        cache.delete(reset_password_cache_key)
+        cache.set(cache_key, hashed_password, ttl=cache_ttl)
```

Comment:
The same way we have `get_cached_reset_password()`  we should have `delete_reset_password()` so that both uses of `ACCOUNT_RESET_PASSWORD_CACHE_KEY` are close to each other